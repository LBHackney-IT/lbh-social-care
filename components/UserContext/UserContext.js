import { createContext, useContext, useEffect, useState } from 'react';
import { useRouter } from 'next/router';

import { AUTH_WHITELIST } from 'utils/auth';
import { getUser } from 'utils/api/checkAuth';
import Layout from 'components/Layout';
import Spinner from 'components/Spinner/Spinner';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState();
  const [redirect, setRedirect] = useState();
  const [isLoading, setLoading] = useState(true);
  const router = useRouter();
  useEffect(() => {
    const initializeAuth = async () => {
      try {
        const data = getUser();
        setUser(data);
      } catch (e) {
        setUser();
        setRedirect(e.response.status === 403 ? '/access-denied' : '/login');
      }
      setLoading(false);
    };
    initializeAuth();
  }, []);
  if (isLoading || !router.isReady) {
    return (
      <Layout>
        <Spinner />
      </Layout>
    );
  }
  if (
    typeof window !== 'undefined' &&
    AUTH_WHITELIST.includes(router.pathname) === Boolean(user)
  ) {
    router.push(redirect || location);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
        isLoading,
        setUser,
        redirect,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
