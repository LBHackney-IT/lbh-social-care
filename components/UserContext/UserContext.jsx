import { createContext, useContext } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children, user }) => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
