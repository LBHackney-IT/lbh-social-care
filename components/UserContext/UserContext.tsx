import { createContext, useContext, ReactNode, ReactElement } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

import type { User } from 'types';

interface Props {
  children: ReactNode;
  user?: User;
}

export const UserContext = createContext<{ user?: User }>({
  user: undefined,
});

export const AuthProvider = ({ children, user }: Props): ReactElement => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export const useAuth = (): { user?: User } => {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
};
