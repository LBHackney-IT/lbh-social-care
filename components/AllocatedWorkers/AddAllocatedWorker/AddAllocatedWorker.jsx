import { useEffect, useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import { useForm } from 'react-hook-form';

import { Radios } from 'components/Form';
import Button from 'components/Button/Button';
import Modal from 'components/Modal/Modal';
import Spinner from 'components/Spinner/Spinner';
import ErrorMessage from 'components/ErrorMessage/ErrorMessage';
import { useAuth } from 'components/UserContext/UserContext';
import {
  getTeams,
  getTeamWorkers,
  addAllocatedWorker,
} from 'utils/api/allocatedWorkers';

const AddAllocatedWorker = ({
  personId,
  currentlyAllocated,
  onAddNewAllocation,
}) => {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [workersLoading, setWorkersLoading] = useState(false);
  const [teams, setTeams] = useState();
  const [workers, setWorkers] = useState();
  const [error, setError] = useState();
  const [postError, setPostError] = useState();
  const [postLoading, setPostLoading] = useState();
  const { handleSubmit, register, watch } = useForm({
    mode: 'onChange',
  });
  const { user } = useAuth();
  const getAllocatedTeams = useCallback(async () => {
    try {
      const data = await getTeams();
      setTeams(data.teams);
    } catch (e) {
      setError(true);
    }
  });
  useEffect(() => {
    getAllocatedTeams();
  }, []);
  const formValues = watch();
  const getAllocatedWorkers = useCallback(async (selectedTeam) => {
    setWorkersLoading(true);
    try {
      const data = await getTeamWorkers(selectedTeam);
      setWorkers(data.workers);
    } catch (e) {
      setError(true);
    }
    setWorkersLoading();
  });
  useEffect(() => {
    formValues.team && getAllocatedWorkers(formValues.team);
  }, [formValues.team]);
  useEffect(() => {
    setPostError();
  }, [formValues.team, formValues.worker]);
  const closeModal = useCallback(() => {
    setIsModalOpen(false);
    setWorkers();
  });
  const addWorker = useCallback(async ({ worker }) => {
    setPostLoading(true);
    setPostError();
    try {
      await addAllocatedWorker(personId, {
        allocatedBy: user.email,
        allocatedWorkerId: worker,
      });
      onAddNewAllocation();
      closeModal();
    } catch (e) {
      setPostError(true);
    }
    setPostLoading(false);
  });
  return (
    <>
      <div className="lbh-table-header">
        <h3 className="govuk-fieldset__legend--m govuk-custom-text-color govuk-!-margin-top-0">
          ALLOCATED WORKER {currentlyAllocated + 1}
        </h3>
        <Button
          label="Allocate worker"
          isSecondary
          onClick={() => setIsModalOpen(true)}
        />
      </div>
      <hr className="govuk-divider" />
      <p>
        <i>{currentlyAllocated === 0 ? 'Currently unallocated' : 'Optional'}</i>
      </p>
      <Modal isOpen={isModalOpen} onRequestClose={closeModal}>
        <h1>Allocate worker</h1>
        <form role="form" onSubmit={handleSubmit(addWorker)}>
          {error ? (
            <ErrorMessage label="Oops an error occurred" />
          ) : (
            <>
              {teams && (
                <Radios
                  name="team"
                  label="Select a team to view workers for that team"
                  labelSize="m"
                  options={teams.map(({ id, name }) => ({
                    value: id,
                    text: name,
                  }))}
                  register={register({ required: true })}
                />
              )}
              {workers && (
                <>
                  <h2>
                    {
                      teams.find(({ id }) => id.toString() === formValues.team)
                        ?.name
                    }{' '}
                    workers
                  </h2>
                  <p className="govuk-body">
                    Select a worker to allocate a case to them. You can view
                    more about a workerâ€™s current allocated cases by selecting
                    their name.
                  </p>
                  <table className="govuk-table">
                    <thead className="govuk-table__head">
                      <tr className="govuk-table__row">
                        <th
                          scope="col"
                          className="govuk-table__header"
                          style={{ width: '4rem' }}
                        >
                          Select
                        </th>
                        <th scope="col" className="govuk-table__header">
                          Name
                        </th>
                        <th
                          scope="col"
                          className="govuk-table__header govuk-table__header--numeric"
                        >
                          Total Cases
                        </th>
                      </tr>
                    </thead>
                    <tbody className="govuk-table__body">
                      {workers.map(
                        ({ firstName, lastName, id, allocationCount }) => (
                          <tr
                            className="govuk-table__row govuk-radios"
                            key={id}
                          >
                            <td className="govuk-table__cell">
                              <div className="govuk-radios__item">
                                <input
                                  aria-labelledby={`worker_${id}`}
                                  className="govuk-radios__input"
                                  name="worker"
                                  type="radio"
                                  value={id}
                                  disabled={workersLoading}
                                  ref={register}
                                />
                                <label className="govuk-label govuk-radios__label"></label>
                              </div>
                            </td>
                            <td className="govuk-table__cell lbh-table__cell--va-middle">
                              <a
                                id={`worker_${id}`}
                                href={`/workers/${id}`}
                                className="govuk-link"
                                rel="noopener"
                                target="_blank"
                              >
                                {firstName} {lastName}
                              </a>
                            </td>
                            <td className="govuk-table__cell govuk-table__cell--numeric">
                              {allocationCount}
                            </td>
                          </tr>
                        )
                      )}
                    </tbody>
                  </table>
                  <Button
                    label="Allocate worker"
                    type="submit"
                    disabled={postLoading}
                  />
                  {postError && (
                    <ErrorMessage label="There was an error allocating the worker. Please retry." />
                  )}
                </>
              )}
              {formValues.team && (!teams || !workers) && <Spinner />}
            </>
          )}
        </form>
      </Modal>
    </>
  );
};

AddAllocatedWorker.propTypes = {
  personId: PropTypes.string.isRequired,
  currentlyAllocated: PropTypes.number.isRequired,
  onAddNewAllocation: PropTypes.func.isRequired,
};

export default AddAllocatedWorker;
