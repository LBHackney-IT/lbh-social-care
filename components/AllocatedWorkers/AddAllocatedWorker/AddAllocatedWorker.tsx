import React, { useState, useCallback, useEffect } from 'react';
import { useRouter } from 'next/router';
import { useForm } from 'react-hook-form';

import { Autocomplete } from 'components/Form';
import Button from 'components/Button/Button';
import Spinner from 'components/Spinner/Spinner';
import ErrorMessage from 'components/ErrorMessage/ErrorMessage';
import { useAuth } from 'components/UserContext/UserContext';
import {
  useTeams,
  useTeamWorkers,
  addAllocatedWorker,
} from 'utils/api/allocatedWorkers';
import { AgeContext, User } from 'types';

interface Props {
  personId: number;
  ageContext: AgeContext;
}

const AddAllocatedWorker = ({
  personId,
  ageContext,
}: Props): React.ReactElement => {
  const [postError, setPostError] = useState<boolean | null>();
  const [postLoading, setPostLoading] = useState<boolean>();
  const { query, push, replace, pathname } = useRouter();
  const { handleSubmit, register, control } = useForm({
    defaultValues: query,
  });
  const { teamId } = query as { teamId?: number };
  const { user } = useAuth() as { user: User };
  const { data: { teams } = {}, error: errorTeams } = useTeams({ ageContext });
  const { data: { workers } = {}, error: errorWorkers } = useTeamWorkers(
    teamId
  );
  const addWorker = useCallback(
    async ({ workerId }) => {
      setPostLoading(true);
      setPostError(null);
      try {
        await addAllocatedWorker(personId, {
          allocatedBy: user.email,
          allocatedTeamId: Number(teamId),
          allocatedWorkerId: Number(workerId),
        });
        push(`/people/${personId}`);
      } catch (e) {
        setPostError(true);
      }
      setPostLoading(false);
    },
    [personId, teamId, push, user.email]
  );
  useEffect(() => {
    setPostError(null);
  }, [query]);
  if (errorTeams || errorWorkers) {
    return <ErrorMessage label="Oops an error occurred" />;
  }
  if (!teams) {
    return <Spinner />;
  }
  return (
    <form role="form" onSubmit={handleSubmit(addWorker)}>
      {teams && (
        <Autocomplete
          name="teamId"
          label="Select a team to view workers for that team"
          labelSize="m"
          placeholder="Select or type team name"
          control={control}
          options={teams.map(({ id, name }) => ({
            value: id,
            text: name,
          }))}
          onChange={(value: number) =>
            replace(
              {
                pathname,
                query: { id: personId, teamId: value },
              },
              undefined,
              { scroll: false }
            )
          }
        />
      )}
      {workers && (
        <>
          <h2>{teams?.find(({ id }) => id === teamId)?.name} workers</h2>
          {workers?.length > 0 ? (
            <>
              <p className="govuk-body">
                Select a worker to allocate a case to them. You can view more
                about a workerâ€™s current allocated cases by selecting their
                name.
              </p>
              <table className="govuk-table">
                <thead className="govuk-table__head">
                  <tr className="govuk-table__row">
                    <th
                      scope="col"
                      className="govuk-table__header"
                      style={{ width: '4rem' }}
                    >
                      Select
                    </th>
                    <th scope="col" className="govuk-table__header">
                      Name
                    </th>
                    <th
                      scope="col"
                      className="govuk-table__header govuk-table__header--numeric"
                    >
                      Total Cases
                    </th>
                  </tr>
                </thead>
                <tbody className="govuk-table__body">
                  {workers.map(
                    ({ firstName, lastName, id, allocationCount }) => (
                      <tr className="govuk-table__row govuk-radios" key={id}>
                        <td className="govuk-table__cell">
                          <div className="govuk-radios__item">
                            <input
                              aria-labelledby={`worker_${id}`}
                              className="govuk-radios__input"
                              name="workerId"
                              type="radio"
                              value={id}
                              disabled={!workers}
                              ref={register}
                              onChange={(e) =>
                                replace(
                                  {
                                    pathname,
                                    query: {
                                      id: personId,
                                      ...query,
                                      workerId: e.target.value,
                                    },
                                  },
                                  undefined,
                                  { scroll: false }
                                )
                              }
                            />
                            <label className="govuk-label govuk-radios__label"></label>
                          </div>
                        </td>
                        <td className="govuk-table__cell lbh-table__cell--va-middle">
                          <a
                            id={`worker_${id}`}
                            href={`/workers/${id}`}
                            className="govuk-link"
                          >
                            {firstName} {lastName}
                          </a>
                        </td>
                        <td className="govuk-table__cell govuk-table__cell--numeric">
                          {allocationCount}
                        </td>
                      </tr>
                    )
                  )}
                </tbody>
              </table>
              <Button
                label="Allocate worker"
                type="submit"
                disabled={postLoading}
              />
            </>
          ) : (
            <p className="govuk-body">Workers not found</p>
          )}
          {postError && (
            <ErrorMessage label="There was an error allocating the worker. Please retry." />
          )}
        </>
      )}
      {teamId && (!teams || !workers) && <Spinner />}
    </form>
  );
};

export default AddAllocatedWorker;
