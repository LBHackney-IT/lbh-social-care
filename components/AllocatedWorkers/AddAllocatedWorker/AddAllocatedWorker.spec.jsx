import { act, fireEvent, render } from '@testing-library/react';

import { UserContext } from 'components/UserContext/UserContext';
import AddAllocatedWorker from './AddAllocatedWorker';

import {
  getTeams,
  getTeamWorkers,
  addAllocatedWorker,
} from 'utils/api/allocatedWorkers';

jest.mock('next/router', () => ({
  useRouter: () => ({
    asPath: 'path',
    push: jest.fn(),
    replace: jest.fn(),
  }),
}));

jest.mock('components/Spinner/Spinner', () => () => 'MockedSpinner');

jest.mock('utils/api/allocatedWorkers', () => ({
  getTeams: jest.fn(),
  getTeamWorkers: jest.fn(),
  addAllocatedWorker: jest.fn(),
}));

describe(`AddAllocatedWorker`, () => {
  getTeams.mockImplementation(() => ({
    data: {
      teams: [
        { id: '1', name: 'Team 1' },
        { id: '2', name: 'Team 2' },
        { id: '3', name: 'Team 3' },
      ],
    },
  }));
  getTeamWorkers.mockImplementation(() => ({
    data: {
      workers: [
        { id: 'a', firstName: 'Worker', lastName: 'A', allocationCount: 3 },
        { id: 'b', firstName: 'Worker', lastName: 'B' },
        { id: 'c', firstName: 'Worker', lastName: 'C' },
      ],
    },
  }));
  const props = {
    personId: '123',
  };

  it('should render properly', async () => {
    const { findByText, getByText, asFragment } = render(
      <UserContext.Provider
        value={{
          user: { name: 'foo', email: 'foo@bar.com' },
        }}
      >
        <AddAllocatedWorker {...props} type="people" />
      </UserContext.Provider>
    );
    const teamRadio = await findByText(
      'Select a team to view workers for that team'
    );
    expect(teamRadio).toBeInTheDocument();
    const radioSelection = getByText('Team 1');
    await act(async () => {
      fireEvent.click(radioSelection);
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should post correctly properly to the API', async () => {
    const { findByText, getByLabelText, getByRole } = render(
      <UserContext.Provider
        value={{
          user: { name: 'foo', email: 'foo@bar.com' },
        }}
      >
        <AddAllocatedWorker {...props} type="people" />
      </UserContext.Provider>
    );
    await findByText('Select a team to view workers for that team');
    await act(async () => {
      fireEvent.click(getByLabelText('Team 3'));
    });
    fireEvent.click(getByLabelText('Worker C'));
    await act(async () => {
      fireEvent.submit(getByRole('form'));
    });
    expect(addAllocatedWorker).toHaveBeenCalled();
    expect(addAllocatedWorker).toHaveBeenCalledWith('123', {
      allocatedBy: 'foo@bar.com',
      allocatedWorkerId: 'c',
    });
  });
});
