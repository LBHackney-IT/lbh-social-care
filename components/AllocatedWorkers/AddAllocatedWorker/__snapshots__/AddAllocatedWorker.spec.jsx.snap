// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAllocatedWorker should render properly 1`] = `
<DocumentFragment>
  <form
    role="form"
  >
    <div
      class="govuk-form-group govuk-grid-column-undefined"
    >
      <label
        class="govuk-label govuk-label--m"
        for="teamId"
      >
        Select a team to view workers for that team 
      </label>
      <div
        aria-expanded="true"
        aria-haspopup="listbox"
        aria-labelledby="downshift-0-label"
        aria-owns="downshift-0-menu"
        class="inputContainer"
        role="combobox"
      >
        <input
          aria-autocomplete="list"
          aria-controls="downshift-0-menu"
          aria-labelledby="downshift-0-label"
          autocomplete="off"
          class="govuk-input govuk-input--width-20"
          data-testid="teamId"
          id="teamId"
          name="teamId.autocomplete"
          placeholder="Select or type team name"
          style="border-color: #000;"
          value=""
        />
      </div>
      <ul
        aria-labelledby="downshift-0-label"
        class="suggestionList"
        id="downshift-0-menu"
        role="listbox"
        style="border-bottom: 1px solid #aaa;"
      >
        <li
          aria-selected="false"
          class="suggestion"
          data-testid="teamId_0"
          id="downshift-0-item-0"
          role="option"
          style="color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-weight: normal;"
        >
          Team 1
        </li>
        <li
          aria-selected="false"
          class="suggestion"
          data-testid="teamId_1"
          id="downshift-0-item-1"
          role="option"
          style="color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-weight: normal;"
        >
          Team 2
        </li>
        <li
          aria-selected="false"
          class="suggestion"
          data-testid="teamId_2"
          id="downshift-0-item-2"
          role="option"
          style="color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-weight: normal;"
        >
          Team 3
        </li>
      </ul>
    </div>
    <h2>
      Team 3 workers
    </h2>
    <p
      class="govuk-body"
    >
      Select a worker to allocate a case to them. You can view more about a workerâ€™s current allocated cases by selecting their name.
    </p>
    <table
      class="govuk-table"
    >
      <thead
        class="govuk-table__head"
      >
        <tr
          class="govuk-table__row"
        >
          <th
            class="govuk-table__header"
            scope="col"
            style="width: 4rem;"
          >
            Select
          </th>
          <th
            class="govuk-table__header"
            scope="col"
          >
            Name
          </th>
          <th
            class="govuk-table__header govuk-table__header--numeric"
            scope="col"
          >
            Total Cases
          </th>
        </tr>
      </thead>
      <tbody
        class="govuk-table__body"
      >
        <tr
          class="govuk-table__row govuk-radios"
        >
          <td
            class="govuk-table__cell"
          >
            <div
              class="govuk-radios__item"
            >
              <input
                aria-labelledby="worker_a"
                class="govuk-radios__input"
                name="workerId"
                type="radio"
                value="a"
              />
              <label
                class="govuk-label govuk-radios__label"
              />
            </div>
          </td>
          <td
            class="govuk-table__cell lbh-table__cell--va-middle"
          >
            <a
              class="govuk-link"
              href="/workers/a"
              id="worker_a"
            >
              Worker A
            </a>
          </td>
          <td
            class="govuk-table__cell govuk-table__cell--numeric"
          >
            3
          </td>
        </tr>
        <tr
          class="govuk-table__row govuk-radios"
        >
          <td
            class="govuk-table__cell"
          >
            <div
              class="govuk-radios__item"
            >
              <input
                aria-labelledby="worker_b"
                class="govuk-radios__input"
                name="workerId"
                type="radio"
                value="b"
              />
              <label
                class="govuk-label govuk-radios__label"
              />
            </div>
          </td>
          <td
            class="govuk-table__cell lbh-table__cell--va-middle"
          >
            <a
              class="govuk-link"
              href="/workers/b"
              id="worker_b"
            >
              Worker B
            </a>
          </td>
          <td
            class="govuk-table__cell govuk-table__cell--numeric"
          />
        </tr>
        <tr
          class="govuk-table__row govuk-radios"
        >
          <td
            class="govuk-table__cell"
          >
            <div
              class="govuk-radios__item"
            >
              <input
                aria-labelledby="worker_c"
                class="govuk-radios__input"
                name="workerId"
                type="radio"
                value="c"
              />
              <label
                class="govuk-label govuk-radios__label"
              />
            </div>
          </td>
          <td
            class="govuk-table__cell lbh-table__cell--va-middle"
          >
            <a
              class="govuk-link"
              href="/workers/c"
              id="worker_c"
            >
              Worker C
            </a>
          </td>
          <td
            class="govuk-table__cell govuk-table__cell--numeric"
          />
        </tr>
      </tbody>
    </table>
    <button
      class="govuk-button"
      data-module="govuk-button"
      type="submit"
    >
      Allocate worker
    </button>
  </form>
</DocumentFragment>
`;
